subsys/bluetooth/audio/ascs.c

```diff
@@ -0,0 +1,1961 @@
+		}
+
+		BT_DBG("ase 0x%02x meta->len %u", meta->ase, meta->len);
+
+		ase = ase_find(ascs, meta->ase);
+		if (!ase) {
+			ascs_cp_rsp_add(meta->ase, BT_ASCS_METADATA_OP,
+					BT_ASCS_RSP_INVALID_ASE, 0x00);
+			BT_ERR("Unable to find ASE");
+			continue;
+		}
+
+		ase_metadata(ase, BT_ASCS_METADATA_OP, meta, buf);
+	}
+
+	return buf->size;
+}
+
+static ssize_t ascs_release(struct bt_ascs *ascs, struct net_buf_simple *buf)
+{
+	const struct bt_ascs_release_op *req;
+	int i;
+
+	if (buf->len < sizeof(*req)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	BT_DBG("num_ases %u", req->num_ases);
+
+	if (buf->len < req->num_ases) {
+		BT_ERR("Malformed ASE Release: len %u < %u", buf->len,
+		       req->num_ases);
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	for (i = 0; i < req->num_ases; i++) {
+		uint8_t id;
+		struct bt_ascs_ase *ase;
+
+		id = net_buf_simple_pull_u8(buf);
+
+		BT_DBG("ase 0x%02x", id);
+
+		ase = ase_find(ascs, id);
+		if (!ase) {
+			ascs_cp_rsp_add(id, BT_ASCS_RELEASE_OP,
+					BT_ASCS_RSP_INVALID_ASE, 0);
+			BT_ERR("Unable to find ASE");
+			continue;
+		}
+
+		ase_release(ase, false);
+	}
+
+	return buf->size;
+}
+
+static ssize_t ascs_cp_write(struct bt_conn *conn,
+			     const struct bt_gatt_attr *attr, const void *data,
+			     uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct bt_ascs *ascs;
+	const struct bt_ascs_ase_cp *req;
+	struct net_buf_simple buf;
+	ssize_t ret;
+
+	if (offset) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len < sizeof(*req)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *) data, len);
+
+	req = net_buf_simple_pull_mem(&buf, sizeof(*req));
+
+	BT_DBG("conn %p attr %p buf %p len %u offset %u op %s (0x%02x)", conn,
+	       attr, data, len, offset, bt_ascs_op_str(req->op), req->op);
+
+	/* Reset/empty response buffer before using it again */
+	net_buf_simple_reset(&rsp_buf);
+
+	ascs = ascs_get(conn);
+	if (!ascs) {
+		BT_ERR("Unable to get ASCS session");
+		len = BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+		ascs_cp_rsp_add(0, req->op, BT_ASCS_RSP_UNSPECIFIED, 0x00);
+		goto respond;
+	}
+
+	switch (req->op) {
+	case BT_ASCS_CONFIG_OP:
+		ret = ascs_config(ascs, &buf);
+		break;
+	case BT_ASCS_QOS_OP:
+		ret = ascs_qos(ascs, &buf);
+		break;
+	case BT_ASCS_ENABLE_OP:
+		ret = ascs_enable(ascs, &buf);
+		break;
+	case BT_ASCS_START_OP:
+		ret = ascs_start(ascs, &buf);
+		break;
+	case BT_ASCS_DISABLE_OP:
+		ret = ascs_disable(ascs, &buf);
+		break;
+	case BT_ASCS_STOP_OP:
+		ret = ascs_stop(ascs, &buf);
+		break;
+	case BT_ASCS_METADATA_OP:
+		ret = ascs_metadata(ascs, &buf);
+		break;
+	case BT_ASCS_RELEASE_OP:
+		ret = ascs_release(ascs, &buf);
+		break;
+	default:
+		ascs_cp_rsp_add(0x00, req->op, BT_ASCS_RSP_NOT_SUPPORTED, 0);
+		BT_DBG("Unknown opcode");
+		len = BT_GATT_ERR(BT_ATT_ERR_NOT_SUPPORTED);
```

Comment:
Does it mean ASE CP notification will be sent as well along with ATT error (as write response)?

Reply:
Ah, I guess it's because ATT Write Without Response might be used. Isn't it?

Reply:
> Ah, I guess it's because ATT Write Without Response might be used. Isn't it?

Correct :)